File: Python/Divide Two Integers.py
"""
Divide two integers without using multiplication, division and mod operator.
If it is overflow, return MAX_INT.
"""

class Solution:
    # @return an integer
    def divide(self, dividend, divisor):
        result, dvd, dvs = 0, abs(dividend), abs(divisor)
        while dvd >= dvs:
            inc = dvs
            i = 0
            while dvd >= inc:
                dvd -= inc
                result += 1 << i
                inc <<= 1
                i += 1
        if dividend > 0 and divisor < 0 or dividend < 0 and divisor > 0:
            if result > 2147483648:
                result = 2147483648
            return -result
        else:
            if result > 2147483647:
                result = 2147483647
            return result

File: Java/Divide Two Integers.java
/*
Divide two integers without using multiplication, division and mod operator.

If it is overflow, return MAX_INT.
*/

/* O(log n) times, O(1) space */
/*
a = (2^n * an + ... + 2^1 * a1 + 2^0 * a0)
We calculate the result of 2^n * an divide by divisor and sum up all the result.
Similarly, we transfer the number into a new number based on divisor:
a = (d^n * an + ... + d^1 * a1 + d^0 * a0)
And then sum up all the an*d^n to get the result.
*/
public class Solution {
    public int divide(int dividend, int divisor) {
        int result = 0;
        
        /* this method cannot give right answer for this situation */
        if (dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }
        
        long dvd = Math.abs((long) dividend), dvs = Math.abs((long) divisor);
        boolean sign = (dividend > 0 && divisor < 0) || (dividend < 0 && divisor > 0);
        
        while (dvd >= dvs) {
            int count = 1;
            long dd = dvs;
            
            while (dvd >= dd) {
                dvd -= dd;
                result += count;
                
                if (dd < (Integer.MAX_VALUE >> 1)) {
                    dd += dd;
                    count += count;
                }
            }
        }
        
        if (sign) return -1 * result;
        return result;
    }
}
